### PR TITLE
RESKC-240: Verify that the Instrumentation Configuration is correct

### PR DESCRIPTION
**Pull request by leoherbie**
DO NOT MERGE UNTIL WE FIX OUR ENVIRONMENTS IN PUPPET.

I recently noticed that the instrumentation configuration is not present in any of our KualiCo environments.  EclipseLink requires instrumentation in order to use features like Lazy.

It used to be obvious when instrumentation was missing because a hard error would occur on application startup.  Then rice switched to static build-time instrumentation which made the error go away.  The error would show up relating to classloading the PersistableBusinessObjectBase class before it was instrumented.  Now that rice is doing static build-time instrumentation this isn't an issue.  What is still an issue is all of our JPA annotated classes.

I briefly looked into using static build time instrumentation but in order to do that you have to list all of your JPA annotated classes in the persistence.xml file.  We don't want to do that.

To solve this problem, I am detecting whether a form of instrumentation is present using our KcConfigVerifier class and throwing an Exception if it is not.

BTW.  This also explains why some devs could not reproduce the issue related to Lazy loading Attachment data but the issue was always present in our KualiCo environments.  It also explains why this issue showed up after we move to KualiCo.  Things were properly configured in previous kuali.org environments.

This fix will also help on prem implementers avoid this subtle nasty problem avoiding erroneous bug reports.

### CODE REVIEW

**Comment by gathreya** [comment]
let me know when this can be merged

**Comment by leoherbie** [comment]
will do

**Comment by leoherbie** [comment]
@gathreya please code review and lets merge in as soon as we can

File: coeus-impl/src/main/java/org/kuali/coeus/sys/impl/KcConfigVerifier.java [hunk 1]
```diff
@@ -61,6 +68,115 @@ protected void verifyOracleConfiguration() {
...
+     * Checks if the Spring Instrumenting ClassLoader is in the ClassLoader chain. If so, this means
+     * that it is properly configured.
+     */
+    protected boolean tomcatInstrumentingClassLoaderConfigured() {
+        //generally this means that the tomcat context.xml file has tomcat ClassLoader configured.  If not then
+        //something really strange is going on
+        ClassLoader cl = this.getClass().getClassLoader();
+        while (cl != null) {
+            if (TOMCAT_INSTRUMENTATION_CLASS_LOADER_CLASS_NAME.equals(cl.getClass().getName())) {
+                return true;
```
Comment (gathreya) [inline coeus-impl/src/main/java/org/kuali/coeus/sys/impl/KcConfigVerifier.java]: do you need to check the version number as well?

Comment (leoherbie) [inline coeus-impl/src/main/java/org/kuali/coeus/sys/impl/KcConfigVerifier.java]: new version of what?  Spring?

Comment (gathreya) [inline coeus-impl/src/main/java/org/kuali/coeus/sys/impl/KcConfigVerifier.java]: of the instrumentation jar.

Comment (leoherbie) [inline coeus-impl/src/main/java/org/kuali/coeus/sys/impl/KcConfigVerifier.java]: I have confirmed that the class files have not been repackaged or anything in newer versions of spring.  beyond that we will have to test this on spring upgrade.  At a quick glance it looks like we will be fine.

**Pull request merged**